tensorboard/plugins/debugger/debugger_plugin.py

```diff
@@ -0,0 +1,504 @@
+from __future__ import division
+from __future__ import print_function
+
+import collections
+import glob
+import json
+import os
+import re
+import threading
+
+import tensorflow as tf
+from werkzeug import wrappers
+
+from tensorboard.backend import http_util
+from tensorboard.backend.event_processing import event_file_loader
+from tensorboard.plugins import base_plugin
+from tensorboard.plugins.debugger import constants
+from tensorboard.plugins.debugger import debugger_server_lib
+
+# HTTP routes.
+_HEALTH_PILLS_ROUTE = '/health_pills'
+_NUMERICS_ALERT_REPORT_ROUTE = '/numerics_alert_report'
+
+# The POST key of HEALTH_PILLS_ROUTE for a JSON list of node names.
+_NODE_NAMES_POST_KEY = 'node_names'
+
+# The POST key of HEALTH_PILLS_ROUTE for the run to retrieve health pills for.
+_RUN_POST_KEY = 'run'
+
+# The default run to retrieve health pills for.
+_DEFAULT_RUN = '.'
+
+# The POST key of HEALTH_PILLS_ROUTE for the specific step to retrieve health
+# pills for.
+_STEP_POST_KEY = 'step'
+
+# A glob pattern for files containing debugger-related events.
+_DEBUGGER_EVENTS_GLOB_PATTERN = 'events.debugger*'
+
+# Encapsulates data for a single health pill.
+HealthPillEvent = collections.namedtuple('HealthPillEvent', [
+    'wall_time', 'step', 'device_name', 'output_slot', 'node_name', 'dtype',
+    'shape', 'value'
+])
+
+
+class DebuggerPlugin(base_plugin.TBPlugin):
+  """TensorFlow Debugger plugin. Receives requests for debugger-related data.
+
+  That data could include health pills, which unveil the status of tensor
+  values.
+  """
+
+  # This string field is used by TensorBoard to generate the paths for routes
+  # provided by this plugin. It must thus be URL-friendly. This field is also
+  # used to uniquely identify this plugin throughout TensorBoard. See BasePlugin
+  # for details.
+  plugin_name = constants.DEBUGGER_PLUGIN_NAME
+
+  def __init__(self, context, debugger_data_server_grpc_port=None):
+    """Constructs a debugger plugin for TensorBoard.
+
+    This plugin adds handlers for retrieving debugger-related data. The plugin
+    also starts a debugger data server once the log directory is passed to the
+    plugin via the call to get_plugin_apps.
+
+    Args:
+      context: A base_plugin.TBContext instance.
+      debugger_data_server_grpc_port: The port at which the debugger data server
+        should listen to gRPC messages. If None, no debugger data server will be
+        started.
+    """
+    self._event_multiplexer = context.multiplexer
+    self._logdir = context.logdir
+    self._debugger_data_server = None
+    self._debugger_data_server_grpc_port = debugger_data_server_grpc_port
+
+    # Start a debugger data server if a port is specified.
+    if self._debugger_data_server_grpc_port:
+      tf.logging.info('Creating DebuggerDataServer at port %d and logdir %s',
+                      self._debugger_data_server_grpc_port, self._logdir)
+      self._debugger_data_server = debugger_server_lib.DebuggerDataServer(
```

Comment:
cc @chihuahua

I share your concern about starting some potentially heavy-weight thing in the constructor. However, I need some help understanding how what you said can be done. In main.py, `create_tb_app` takes a list of plugin constructors, instead of a list of actual plugin objects. So there is no way to call the proposed `listen` method in main.py. Maybe modify ``standard_tensorboard_wsgi` in `application.py` so it can call the `listen` method? But then there is the problem of how to pass the port argument in. Maybe through a flag? That feels a little messy. Let me know if there is a cleaner way to do it.